Refactor custom op tests

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31282 Refactor custom op tests**

Introduce a helper to easily call stack ops

Differential Revision: [D19061515](https://our.internmc.facebook.com/intern/diff/D19061515/)

smessmer committed:
Refactor custom op tests

Introduce a helper to easily call stack ops

Differential Revision: [D19061515](https://our.internmc.facebook.com/intern/diff/D19061515/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor custom op tests"


Introduce a helper to easily call stack ops

Differential Revision: [D19061515](https://our.internmc.facebook.com/intern/diff/D19061515/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit e49d95d:
* **3/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e49d95d97b46a0d0f0ef58b556a1fc84a63e8836).

### 3 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3973809) pytorch_linux_xenial_py3_6_gcc5_4_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75639989) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75639989))
<details>
<summary>
<code>Dec 17 02:16:34 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_operator_from_registry_and_execute()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:42:80: error: conversion from \'void\' to non-scalar type \'std::vector<at::Tensor>\' requested\n     helpers::get_operator_from_registry_and_execute<std::vector<torch::Tensor>>(\"custom::op\", torch::ones(5), 2.0, 3);\n                                                                                ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_autograd_operator_from_registry_and_execute()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:67: error: conversion from \'void\' to non-scalar type \'at::Tensor\' requested\n     helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n                                                                   ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_autograd_operator_from_registry_and_execute_in_nograd_mode()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:75:67: error: conversion from \'void\' to non-scalar type \'at::Tensor\' requested\n     helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n                                                                   ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = std::vector<at::Tensor>; Args = {at::Tensor, double, int}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:42:117:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n   return torch::jit::pop(stack).to<Result>();\n                                            ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = at::Tensor; Args = {at::Tensor&, int, at::Tensor&}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:103:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n" }</code>
</summary>

```
Dec 17 02:16:34 CMakeFiles/test_custom_ops.dir/build.make:62: recipe for target 'CMakeFiles/test_custom_ops.dir/test_custom_ops.cpp.o' failed 
Dec 17 02:16:34 make[2]: Leaving directory '/var/lib/jenkins/custom-op-build' 
Dec 17 02:16:34 CMakeFiles/Makefile2:141: recipe for target 'CMakeFiles/test_custom_ops.dir/all' failed 
Dec 17 02:16:34 make[1]: Leaving directory '/var/lib/jenkins/custom-op-build' 
Dec 17 02:16:34 make: *** [all] Error 2 
Dec 17 02:16:34 Makefile:83: recipe for target 'all' failed 
Dec 17 02:16:34 + cleanup 
Dec 17 02:16:34 + retcode=2 
Dec 17 02:16:34 + set +x 
Dec 17 02:16:34 =================== sccache compilation log =================== 
ar/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = at::Tensor; Args = {at::Tensor&, int, at::Tensor&}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:103:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n" } 
Dec 17 02:16:34  
Dec 17 02:16:34 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 17 02:16:34 Compile requests              2510 
Dec 17 02:16:34 Compile requests executed     2252 
Dec 17 02:16:34 Cache hits                    2237 
Dec 17 02:16:34 Cache misses                     0 
Dec 17 02:16:34 Cache timeouts                   0 
Dec 17 02:16:34 Cache read errors                0 
Dec 17 02:16:34 Forced recaches                  0 
Dec 17 02:16:34 Cache write errors               0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3973801) pytorch_xla_linux_xenial_py3_6_clang7_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75641320) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75641320))
<details>
<summary>
<code>Dec 17 02:17:04 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:41:30: error: no viable conversion from \'void\' to \'std::vector<torch::Tensor>\'\n  std::vector<torch::Tensor> output =\n                             ^\n/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_vector.h:318:7: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'const std::vector<at::Tensor, std::allocator<at::Tensor> > &\' for 1st argument\n      vector(const vector& __x)\n      ^\n/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_vector.h:335:7: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'std::vector<at::Tensor, std::allocator<at::Tensor> > &&\' for 1st argument\n      vector(vector&& __x) noexcept\n      ^\n/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_vector.h:373:7: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'initializer_list<std::vector<at::Tensor, std::allocator<at::Tensor> >::value_type>\' (aka \'initializer_list<at::Tensor>\') for 1st argument\n      vector(initializer_list<value_type> __l,\n      ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:57:17: error: no viable conversion from \'void\' to \'torch::Tensor\'\n  torch::Tensor output =\n                ^\n/opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/TensorBody.h:87:3: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'const at::Tensor &\' for 1st argument\n  Tensor(const Tensor&) = default;\n  ^\n/opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/TensorBody.h:88:3: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'at::Tensor &&\' for 1st argument\n  Tensor(Tensor&&) = default;\n  ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:74:17: error: no viable conversion from \'void\' to \'torch::Tensor\'\n  torch::Tensor output =\n                ^\n/opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/TensorBody.h:87:3: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'const at::Tensor &\' for 1st argument\n  Tensor(const Tensor&) = default;\n  ^\n/opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/TensorBody.h:88:3: note: candidate constructor not viable: cannot convert argument of incomplete type \'void\' to \'at::Tensor &&\' for 1st argument\n  Tensor(Tensor&&) = default;\n  ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:3: error: void function \'get_operator_from_registry_and_execute\' should not return a value [-Wreturn-type]\n  return torch::jit::pop(stack).to<Result>();\n  ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:42:14: note: in instantiation of function template specialization \'helpers::get_operator_from_registry_and_execute<std::vector<at::Tensor, std::allocator<at::Tensor> >, at::Tensor, double, int>\' requested here\n    helpers::get_operator_from_registry_and_execute<std::vector<torch::Tensor>>(\"custom::op\", torch::ones(5), 2.0, 3);\n             ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:3: error: void function \'get_operator_from_registry_and_execute\' should not return a value [-Wreturn-type]\n  return torch::jit::pop(stack).to<Result>();\n  ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:14: note: in instantiation of function template specialization \'helpers::get_operator_from_registry_and_execute<at::Tensor, at::Tensor &, int, at::Tensor &>\' requested here\n    helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n             ^\n5 errors generated.\n" }</code>
</summary>

```
Dec 17 02:17:04 CMakeFiles/test_custom_ops.dir/build.make:62: recipe for target 'CMakeFiles/test_custom_ops.dir/test_custom_ops.cpp.o' failed 
Dec 17 02:17:04 make[2]: Leaving directory '/var/lib/jenkins/custom-op-build' 
Dec 17 02:17:04 CMakeFiles/Makefile2:141: recipe for target 'CMakeFiles/test_custom_ops.dir/all' failed 
Dec 17 02:17:04 make[1]: Leaving directory '/var/lib/jenkins/custom-op-build' 
Dec 17 02:17:04 make: *** [all] Error 2 
Dec 17 02:17:04 Makefile:83: recipe for target 'all' failed 
Dec 17 02:17:04 + cleanup 
Dec 17 02:17:04 + retcode=2 
Dec 17 02:17:04 + set +x 
Dec 17 02:17:04 =================== sccache compilation log =================== 
urn-type]\n  return torch::jit::pop(stack).to<Result>();\n  ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:14: note: in instantiation of function template specialization \'helpers::get_operator_from_registry_and_execute<at::Tensor, at::Tensor &, int, at::Tensor &>\' requested here\n    helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n             ^\n5 errors generated.\n" } 
Dec 17 02:17:04  
Dec 17 02:17:04 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 17 02:17:04 Compile requests              2518 
Dec 17 02:17:04 Compile requests executed     2260 
Dec 17 02:17:04 Cache hits                    2246 
Dec 17 02:17:04 Cache misses                     0 
Dec 17 02:17:04 Cache timeouts                   0 
Dec 17 02:17:04 Cache read errors                0 
Dec 17 02:17:04 Forced recaches                  0 
Dec 17 02:17:04 Cache write errors               0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3973802) pytorch_linux_xenial_py2_7_9_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75641755) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75641755))
<details>
<summary>
<code>Dec 17 02:17:14 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_operator_from_registry_and_execute()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:42:80: error: conversion from \'void\' to non-scalar type \'std::vector<at::Tensor>\' requested\n     helpers::get_operator_from_registry_and_execute<std::vector<torch::Tensor>>(\"custom::op\", torch::ones(5), 2.0, 3);\n     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_autograd_operator_from_registry_and_execute()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:67: error: conversion from \'void\' to non-scalar type \'at::Tensor\' requested\n     helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In function \'void get_autograd_operator_from_registry_and_execute_in_nograd_mode()\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:75:67: error: conversion from \'void\' to non-scalar type \'at::Tensor\' requested\n     helpers::get_operator_from_registry_and_execute<torch::Tensor>(\"custom::op_with_autograd\", x, 2, y);\n     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = std::vector<at::Tensor>; Args = {at::Tensor, double, int}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:42:117:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n   return torch::jit::pop(stack).to<Result>();\n                                            ^\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = at::Tensor; Args = {at::Tensor&, int, at::Tensor&}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:103:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n" }</code>
</summary>

```
Dec 17 02:17:14 make[2]: *** [CMakeFiles/test_custom_ops.dir/test_custom_ops.cpp.o] Error 1 
Dec 17 02:17:14 CMakeFiles/Makefile2:141: recipe for target 'CMakeFiles/test_custom_ops.dir/all' failed 
Dec 17 02:17:14 make[1]: Leaving directory '/var/lib/jenkins/custom-op-build' 
Dec 17 02:17:14 Makefile:83: recipe for target 'all' failed 
Dec 17 02:17:14 make[1]: *** [CMakeFiles/test_custom_ops.dir/all] Error 2 
Dec 17 02:17:14 make: *** [all] Error 2 
Dec 17 02:17:14 + cleanup 
Dec 17 02:17:14 + retcode=2 
Dec 17 02:17:14 + set +x 
Dec 17 02:17:14 =================== sccache compilation log =================== 
ar/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp: In instantiation of \'void helpers::get_operator_from_registry_and_execute(const char*, Args&& ...) [with Result = at::Tensor; Args = {at::Tensor&, int, at::Tensor&}]\':\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:58:103:   required from here\n/var/lib/jenkins/workspace/test/custom_operator/test_custom_ops.cpp:36:44: error: return-statement with a value, in function returning \'void\' [-fpermissive]\n" } 
Dec 17 02:17:14  
Dec 17 02:17:14 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 17 02:17:14 Compile requests              2519 
Dec 17 02:17:14 Compile requests executed     2260 
Dec 17 02:17:14 Cache hits                    2246 
Dec 17 02:17:14 Cache misses                     0 
Dec 17 02:17:14 Cache timeouts                   0 
Dec 17 02:17:14 Cache read errors                0 
Dec 17 02:17:14 Forced recaches                  0 
Dec 17 02:17:14 Cache write errors               0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "Refactor custom op tests"


Introduce a helper to easily call stack ops

Differential Revision: [D19061515](https://our.internmc.facebook.com/intern/diff/D19061515/)

[ghstack-poisoned]
